quickwit/quickwit-search/src/collector.rs

```diff
@@ -293,25 +233,64 @@ enum AggregationSegmentCollectors {
 pub struct QuickwitSegmentCollector {
     num_hits: u64,
     split_id: String,
-    score_extractor: SortingFieldExtractorPair,
+    score_extractors: Vec<SortingFieldExtractorComponent>,
     // PartialHits in this heap don't contain a split_id yet.
     top_k_hits: TopK<PartialHit, PartialHitSortingKey, HitSortingMapper>,
     segment_ord: u32,
     timestamp_filter_opt: Option<TimestampFilter>,
     aggregation: Option<AggregationSegmentCollectors>,
+    search_after: Option<PartialHit>,
 }
 
 impl QuickwitSegmentCollector {
     #[inline]
     fn collect_top_k(&mut self, doc_id: DocId, score: Score) {
-        let (sort_value, sort_value2) =
-            self.score_extractor.extract_typed_sort_value(doc_id, score);
+        let sort_values: Vec<_> = self
+            .score_extractors
+            .iter()
+            .map(|extractor| extractor.extract_typed_sort_value_opt(doc_id, score))
+            .collect();
+
+        if let Some(search_after) = &self.search_after {
+            let search_after_values = [
+                search_after.sort_value.and_then(|v| v.sort_value),
+                search_after.sort_value2.and_then(|v| v.sort_value),
+            ];
+            let mut cmp_result = compare_fields(
+                &self.top_k_hits.sort_key_mapper.orders,
+                &sort_values,
+                &search_after_values,
+            );
+            if !search_after.split_id.is_empty() {
+                // TODO actually it's not first, it should be what's in _shard_doc then first then
+                // default
+                let order = self
+                    .top_k_hits
+                    .sort_key_mapper
+                    .orders
+                    .first()
+                    .unwrap_or(&SortOrder::Desc);
+                cmp_result = cmp_result
+                    .then_with(|| order.compare(&self.split_id, &search_after.split_id))
+                    .then_with(|| order.compare(&self.segment_ord, &search_after.segment_ord))
+                    .then_with(|| order.compare(&doc_id, &search_after.doc_id))
+            }
+
+            if cmp_result != Ordering::Less {
+                return;
+            }
+        }
+
+        // TODO better partial hit, again
+        let mut sort_values = sort_values.into_iter();
+        let sort_value = sort_values.next().flatten();
+        let sort_value2 = sort_values.next().flatten();
 
         let hit = PartialHit {
             sort_value: sort_value.map(Into::into),
             sort_value2: sort_value2.map(Into::into),
             // we actually know the split_id, but the hit is likely to be discarded, so clonning it
-            // would cause a probably useless allocation. TODO use an Arc<str> instead?
+            // would cause a probably useless allocation. TODO use an Arc<str>/Rc<str> instead?
```

Comment:
In a different PR (it won't be merged before yours, no worry), I removed those by introducing a SegmentPartialHit https://github.com/quickwit-oss/quickwit/pull/3915/files#diff-e46a7a27f5f472b1b9897b219c56db65c7af89a6ee04d783a402d4688edd4d13

I think it is  not a bad idea. If you agree could we do this refactoring?